### PR TITLE
feat: Auto logout impersonation

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

It's best practice to have a reliable system for downgrading priveleges, for example logging in as a user.

## Changes

* Adds a middleware that sets an expiry time on impersonated sessions and automatically logs them out after a set time period

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## How did you test this code?

Couldn't think how to come up with a reliable test here. I'll give it a go though...

### CODE REVIEW

File: posthog/middleware.py [hunk 1]
```diff
@@ -629,3 +630,39 @@ def process_response(self, request, response):
...
+                    "Impersonation session has expired. Please log in again.",
+                    status=401,
+                )
+            elif not request.path.startswith("/logout"):
+                # Any other endpoint will redirect to the logout page (this could be doc level calls or other things like static assets)
+                return redirect("/logout/")
+            else:
+                # And finally we restore the login and redirect to the admin page
+                restore_original_login(request)
+                return redirect("/admin/")
```
Comment (webjunkie) [inline posthog/middleware.py:661]: What is left here in this else branch? Wouldn't this just match the `/logout` route? And for logout, couldn't we just let the request run through (it would stop impersonating, right)?

File: posthog/middleware.py [hunk 2]
```diff
@@ -629,3 +630,39 @@ def process_response(self, request, response):
...
+
+            if request.path.startswith("/api/"):
+                # If we respond with a 401, the frontend will check and redirect to logout
+                return HttpResponse(
+                    "Impersonation session has expired. Please log in again.",
+                    status=401,
+                )
+            elif not request.path.startswith("/logout"):
+                # Any other endpoint will redirect to the logout page (this could be doc level calls or other things like static assets)
+                return redirect("/logout/")
```
Comment (webjunkie) [inline posthog/middleware.py:657]: I'm wondering if this would log you out completely in case the app makes multiple requests to api and static assets and such.

Comment (benjackwhite) [reply posthog/middleware.py:657]: Yeah this is the problem I'm finding. That and it not working in tests for some reason.

Essentially I want a single reliable way to restore the old session but seeing as this request can be anything (even static assets) its hard to parse what the right way to do it is...
